Ooba charhub

ewof opened the pull request:
fixes ooba adapter
add CHUB tab for downloading memory books and characters
user clicks on an item and it opens up the appropriate edit modal 

known issues idk how to fix:
- when ur a guest and u import a memory book they aren't given an id so u cant visit them but when ur logged in they do get ids

video demonstration on desktop

https://github.com/luminai-companion/agn-ai/assets/26354814/2aed4875-5463-457e-a7b0-83bf5e079674


mobile screenshot
![image](https://user-images.githubusercontent.com/26354814/236653664-0c43841f-ca0c-4835-9d8f-c06abd0439eb.png)

ewof committed:
update to new oob api
fix add bos token in gen settings not having a label
add skip special tokens

ewof committed:
not in api anymore dont need this

ewof committed:
part 1

ewof committed:
use correct avatars url and add left/right buttons

ewof committed:
make buttons work and add page number

ewof committed:
import modal

ewof committed:
make edits in modals work
some renaming
give all delete buttons red schema

ewof committed:
update descriptions

ewof committed:
make this optional

ewof committed:
fix search and remove unused imports

ewof committed:
set page to 1 when u change filter settings

ewof committed:
link to original item on chub in modals

ewof committed:
set page to 1 on search bar change

ewof committed:
extractCardData not extractTavernData

ewof committed:
remove this padding again

ewof committed:
leave it up to user to set version of api

ewof committed:
merge upstream

sceuick commented on srv/adapter/openai.ts:
```diff
@@ -10,7 +10,7 @@ import { AppSchema } from '../db/schema'
 import { DEFAULT_SUMMARY_PROMPT } from '/common/image'
 
-const baseUrl = `https://api.openai.com`
+const baseUrl = `https://api.openai.com/`
```
Needs to be reverted

sceuick commented on web/pages/Chat/ChatDetail.tsx:
```diff
@@ -278,7 +278,7 @@ const ChatDetail: Component = () => {
               </div>
             </Show>
-            <div class="flex flex-col-reverse gap-4 overflow-y-scroll pr-2 sm:pr-4">
+            <div class="flex flex-col-reverse gap-4 overflow-y-scroll">
```
Needs reverting

ewof replied:
:((((((
it looks weird with it

sceuick commented on web/pages/Chub/BookList.tsx:
```diff
@@ -0,0 +1,40 @@
+          )}
+        </For>
+        <Show when={books.length < 4}>
+          <For each={new Array(4 - books.length)}>{() => <div></div>}</For>
```
This should be `books().length` I believe. Otherwise you're return the `.length` of a function which is it's arity.

sceuick commented on web/pages/Chub/BookList.tsx:
```diff
@@ -0,0 +1,40 @@
+              name={book.name}
+              fullPath={book.fullPath}
+              avatar={
+                `https://avatars.charhub.io/avatars/${book.fullPath}/avatar.webp` ||
```
This will never coalesce to the second string so it may as well be removed.

sceuick commented on web/pages/Chub/BookList.tsx:
```diff
@@ -0,0 +1,40 @@
+  })
+
+  const books = createMemo(() => {
+    return chubStore().books
```
Use `const books = chubStore(s => s.books)` instead.

sceuick commented on web/pages/Chub/BookList.tsx:
```diff
@@ -0,0 +1,40 @@
+import ChubNavigation from './ChubNavigation'
+
+const BookList: Component = () => {
+  createEffect(() => {
```
We've been migrating to `onMount`, but this should be fine.

sceuick commented on web/pages/Chub/CharList.tsx:
```diff
@@ -0,0 +1,39 @@
+            />
+          )}
+        </For>
+        <Show when={chars.length < 4}>
```
Same here with `chars().length` if you're referencing a memoized value. If you change `chars` as suggested above then this will be correct.

sceuick commented on web/pages/Chub/CharList.tsx:
```diff
@@ -0,0 +1,39 @@
+  })
+
+  const chars = createMemo(() => {
+    return chubStore().chars
```
This should be `const chars = chubStore(s => s.chars)`

sceuick commented on web/pages/Chub/ChubImportBook.tsx:
```diff
@@ -0,0 +1,92 @@
+              book={book()}
+              entrySort={entrySort()}
+              updateEntrySort={updateEntrySort}
+              onChange={(b) => {
```
`onChange={setBook}` for the shorthand version.

sceuick commented on web/pages/Chub/ChubImportChar.tsx:
```diff
@@ -0,0 +1,137 @@
+            }
+            value={char().name}
+            onChange={(e) => {
+              char().name = e.currentTarget.value
```
Use the `setChar` function to update signals.

sceuick commented on web/pages/Chub/ChubItem.tsx:
```diff
@@ -0,0 +1,114 @@
+
+      const blob = await res.blob()
+
+      processImage(new File([blob], `main_${props.fullPath}.png`))
```
You'll need to set the mime type  here as the 3rd argument: `{ type: 'image/png' }` otherwise it'll be rejected by the Agnai api.

sceuick commented on web/pages/Chub/ChubNavigation.tsx:
```diff
@@ -0,0 +1,84 @@
+            fieldName="search"
+            placeholder="Search by name..."
+            value={chubOptions.search}
+            onKeyUp={(ev) => {
```
Move inline functions like this to outside of the XML to the markup is easier to follow.

sceuick commented on web/pages/Chub/Chub.tsx:
```diff
@@ -0,0 +1,62 @@
+import { ChubOptions } from '../../store/chub'
+import BookList from './BookList'
+
+export const chubOptions: ChubOptions = {
```
This should be in state (if it's used in multiple places) or a signal if it's used in a single place.
I believe we've used local storage elsewhere to cache/re-hydrate these sorts of sort/filtering values.

sceuick commented on web/pages/Chub/ChubNavigation.tsx:
```diff
@@ -0,0 +1,84 @@
+            placeholder="Search by name..."
+            value={chubOptions.search}
+            onKeyUp={(ev) => {
+              chubOptions.search = ev.currentTarget.value
```
We definitely want to avoid these styles of mutations. We should stick with idiomatic state management (chubStore or signals, whichever suits best).

sceuick commented on web/store/chub.ts:
```diff
@@ -0,0 +1,124 @@
+  return {
+    async getChubChars() {
+      let sort = ''
+      switch (chubOptions.sort) {
```
I'd break this out to a top-level function.
The `chubOptions` seems suitably placed in this store.

sceuick reviewed (changes requested):
It may also be worth trying to re-use the cards from the characters list as well.

sceuick reviewed (changes requested):
It may also be worth trying to re-use the cards from the characters list as well.

ewof committed:
inital reviews

ewof committed:
fix setting page number
check for duplicate chars on import

ewof committed:
default to creation date (all sfw image previews)

ewof committed:
fix dupe check

ewof committed:
check for dupe memory books on import

ewof committed:
nevermind

ewof committed:
forgot routes

LukePavelka commented on web/Navigation.tsx:
```diff
@@ -11,6 +11,7 @@ import {
   MailPlus,
   MessageCircle,
   Settings,
+  ShoppingBag,
```
why is it called a ShoppingBag?

ewof replied:
that's the icon for the chub tab, from lucide-solid
![image](https://github.com/agnaistic/agnai/assets/26354814/369a2566-cf2c-49cf-ac19-135a9a386238)

LukePavelka commented on web/shared/GenerationSettings.tsx:
```diff
@@ -441,20 +441,27 @@ const GenSettings: Component<Props> = (props) => {
         service={props.service}
         aiSetting={'banEosToken'}
       />
+      <Toggle
```
Excellent :)

LukePavelka reviewed (approved):
make a squash, it would be good to split it into two commitu. fix textgen + chub. but I understand that's probably not possible anymore.

I'd like to ask, I'm not entirely clear on the code, the main reason for the textgen <not work> is that their api was different than what we called?

ewof committed:
fix

ewof committed:
use chub.ai instead of characterhub.org

sceuick committed:
- remove circular refs
- move page to state
- remove fetches from components
- add feature flag
- add chub subpage routes

sceuick committed:
- Remove direct mutators from memory
- Fix loading positioning

sceuick committed:
Fix force preset modal

sceuick committed:
Fix empty bot background
